### PR TITLE
Basic validation of AJAX responses

### PR DESCRIPTION
**Pull request by swissspidy**
**Examples:**

Internal Server Error because of an infinite loop:

screen shot 2016-06-12 at 16 13 35

Custom `die()` call in the ajax callback:

screen shot 2016-06-12 at 15 59 56

**To do:**

* [x] Change wording depending on action (update, install, etc.) or simply make it more generic

Fixes obenland/shiny-updates#272.

### CODE REVIEW

**Comment by obenland** [comment]
Heads up, a little bit of a questions avalanche :) 

Is there a good way to test this PR? What kinds of errors did you trigger when you wrote it?
There is no way to handle that in `ajaxAlways()` and cancel the error callback, is there?
Do you think we should call the function `isResponseValid()` since it returns a boolean?
The function accounts for update buttons, should it also account for update notices?
Could this also be a good opportunity to handle `-1` responses for failed admin ajax referrers?

File: src/wp-admin/js/updates.js [hunk 1]
```diff
@@ -1730,6 +1758,51 @@
...
+	 * If the response deems to be invalid, an admin notice is being displayed.
+	 *
+	 * @typedef {object} validateResponse
+	 * @param {object} response Response from the server.
+	 */
+	wp.updates.validateResponse = function( response ) {
+		var reason = wp.updates.l10n.updateFailedShort;
+
+		// Make sure the response is a valid data object and not a Promise object.
+		if ( 'object' === typeof response && ! 'function' === typeof response.always ) {
```
Comment (swissspidy) [inline src/wp-admin/js/updates.js]: We can use Underscore's helper functions here, e.g. `_.isObject( response )` etc. See `wp.ajax.send` for am example.

**Comment by ocean90** [comment]
@obenland For a fatal error you could use the plugin mentioned by Jorbin: https://wordpress.slack.com/archives/core/p1465417741000862

> I have a plugin that fatal errors on version 0.2 if you want to check with that.  `wp plugin install this-plugin-should-not-be-used --version=0.1`

**Comment by swissspidy** [comment]
> Is there a good way to test this PR? What kinds of errors did you trigger when you wrote it? (Reading the ticket helps)

Apart from the mentioned plugin, you can also just put something like `die( 'abc' );` or `while(1){}` into `wp_ajax_update_plugin()`.

> There is no way to handle that in ajaxAlways() and cancel the error callback, is there?

I don't think so. `ajaxAlways` is always run, even for successful responses. That means we would have to copy the validation logic inside `wp.ajax.send`.
Also, `ajaxAlways` is run _after_ the callbacks have been called, but we want it _before_ (or inside the callback FWIW).

Since `wp.ajax.send()` uses `deferred.promise()`, we can't reject the Promise if the response is invalid. Per [the docs](https://api.jquery.com/deferred.promise/):

> The deferred.promise() method allows an asynchronous function to prevent other code from interfering with the progress or status of its internal request. The Promise exposes only the Deferred methods needed to attach additional handlers or determine the state (then, done, fail, always, pipe, progress, state and promise), but not ones that change the state (resolve, reject, notify, resolveWith, rejectWith, and notifyWith).

So, we can't really change that without modifying `wp.ajax.send()`.

> Do you think we should call the function isResponseValid() since it returns a boolean?

`isValidResponse()` sounds good if we keep the function that way and do not do something like adding it to `ajaxAlways` or hooking into `wp.ajax.send`. Depends, really.

> The function accounts for update buttons, should it also account for update notices?

Yes. It's mainly a proof-of-concept so far, haven't looked at all screens etc.

> Could this also be a good opportunity to handle -1 responses for failed admin ajax referrers?

Yeah `wp.ajax.send` checks for `1` values already, so it makes sense to check for `-1` as well there.

File: src/wp-admin/js/updates.js [hunk 2]
```diff
@@ -1762,43 +1762,52 @@
...
+		$( '.button.updating-message' )
 			.removeClass( 'updating-message' )
 			.attr( 'aria-label', wp.updates.l10n.updateFailedShort )
 			.prop( 'disabled', true )
 			.text( wp.updates.l10n.updateFailedShort );
 
+		$( '.updating-message:not(.button):not(.thickbox)' )
+			.removeClass( 'updating-message notice-warning' )
+			.addClass( 'notice-error' )
+			.find( 'p' ).html( errorMessage );
```
Comment (swissspidy) [inline src/wp-admin/js/updates.js]: Should probably use `.text()` to escape the errorMessage / response

Comment (obenland) [inline src/wp-admin/js/updates.js]: Good call

**Comment by swissspidy** [comment]
#mergereview @obenland

**Comment by obenland** [comment]
`errorReason` could probably just be `error`. `unknownError` should also be defined in the test fixtures. Looks good besides that, thanks for your work on it!

**Comment by obenland** [comment]
Oh, if you want to add a new line between `break;` and the next `case:` in the switch statement, my OCD would thank you :)

**Pull request merged**